File: Programmars/Level_2/가장 큰 수.py
# 순열(완전탐색) -> 시간 초과
def solution(numbers):
    answer = ''
    numbers = list(map(str, numbers))
    numbers.sort(key=lambda x: x[0:], reverse=True)
    print(numbers)
    first = -1
    temp = []
    for num in numbers:
        print(first)
        if first == num[0]:
            temp.append(num)
        else:
            if temp:
                temp.sort(key=lambda x: x[1:])
                print(temp)
                answer += ''.join(temp)
                temp = []
            else:
                temp.append(num)
            first = num[0]
    if temp:
        temp.sort(key=lambda x: x[1:])
        print(temp)
        answer += ''.join(temp)
    return answer


print(solution([6, 10, 2]))  # "6210"
print(solution([3, 30, 34, 5, 9]))  # "9534330"

print(solution([505, 50, 5]))  # 550550
print(solution([0, 0, 0, 0]))  # 0


# ref: lambda 함수로 문자열 변환
def solution(numbers):
    numbers = list(map(str, numbers))
    numbers.sort(key=lambda x: x*3, reverse=True)
    return str(int(''.join(numbers)))

File: JavaScript/Inflearn/py_4_자료구조 활용/가장 큰 수.js
// 숫자의 자릿수 중 m개 제거해서 가장 큰 수 만들기
// 숫자 순서 유지해야 함
const solution = (n, m) => {
  const numbers = String(n).split('');
  let stack = [];
  numbers.forEach((num) => {
    // ✅ m이 0 이상일 때 다음 수가 스택 마지막 수보다 크면 pop
    while (stack.length > 0 && m > 0 && stack[stack.length - 1] < num) {
      stack.pop();
      m--;
    }
    stack.push(num);
  });

  if (m !== 0) stack = stack.slice(0, stack.length - m);

  return +stack.join('');
};

console.log(solution(5276823, 3)); // 7823
console.log(solution(9977252641, 5)); // 99776
